cdap-source-control/src/main/java/io/cdap/cdap/sourcecontrol/operationrunner/InMemorySourceControlOperationRunner.java

```diff
@@ -155,5 +163,45 @@ private Path validateAppConfigRelativePath(RepositoryManager repositoryManager,
     }
     return filePath;
   }
+
+  @Override
+  public ListAppsResponse list(NamespaceId namespace, RepositoryConfig repoConfig) throws ListAppsFailureException,
+    AuthenticationConfigException {
+    try (RepositoryManager repositoryManager = repoManagerFactory.create(namespace, repoConfig)) {
+      String currentCommit = repositoryManager.cloneRemote();
+      Path basePath = repositoryManager.getBasePath();
+
+      FileFilter configFileFilter = getConfigFileFilter();
+      List<File> configFiles = DirUtils.listFiles(basePath.toFile(), configFileFilter);
+
+      List<ListAppResponse> responses = new ArrayList<>();
+      for (File configFile : configFiles) {
+        String applicationName = FileUtils.getNameWithoutExtension(configFile.getName());
+        String fileHash = repositoryManager.getFileHash(basePath.relativize(configFile.toPath()), currentCommit);
+        responses.add(new ListAppResponse(applicationName, fileHash));
+      }
+
+      // Sorting by application name to make response consistent.
+      responses.sort(Comparator.comparing(ListAppResponse::getName));
+
+      return new ListAppsResponse(responses);
+    } catch (IOException | GitAPIException | NotFoundException e) {
```

Comment:
Is the `NotFoundException` caused by `repositoryManager.getFileHash`? We might be able to remove it for later refactoring

Reply:
Yes it is from `getFileHash`. We can remove it after afterhttps://github.com/cdapio/cdap/pull/url gets merged